docs/installation/production/configuration/advanced/django_channels_layer.rst

```diff
@@ -0,0 +1,96 @@
+********************************
+Django Channels Layer (Optional)
+********************************
+
+**Last Updated:** Sep 2020
+
+Production installations that use the ``WebSockets`` and/or ``Bokeh Server`` functionality that comes with Tethys, require the a ``CHANNEL_LAYER``.
+
+Key Concepts
+============
+A ``CHANNEL_LAYER`` provides a backend where ``WebSocket`` messages can be stored and then accessed by  different app instances (ASGI processes).
+
+Adding a REDIS CHANNEL_LAYER
+============================
+Development installations make use of the default ``InMemoryChannelLayer``, however this layer has some limitations in production (See `Django Channels documentation <https://channels.readthedocs.io/en/latest/topics/channel_layers.html#in-memory-channel-layer>`_). To address these limitations, ``Django Channels`` suppports a ``REDIS CHANNEL LAYER``, however other ``CHANNEL_LAYERS`` can be configured. The following documentation demonstrates how to configure a ``REDIS CHANNEL LAYER``.
+
+First, install the ``channels_redis`` Python package in your Tethys conda environment.
+
+::
+
+    pip install channels_redis
+
+Second, add or modify the ``CHANNEL_LAYERS`` parameter in the ``portal_config.yml`` as follows:
+
+.. code-block:: yaml
+
+    CHANNEL_LAYERS:
+      default:
+        BACKEND: channels_redis.core.RedisChannelLayer
+        CONFIG:
+          hosts:
+          - [127.0.0.1, 6379]
+
+Finally, start a redis instance. This can easily be done with docker as shown below.
+
+::
+
+    docker run -p 6379:6379 -d redis:5
+
+Once a production ``CHANNEL_LAYER`` has been configured, the number of ``ASGI_PROCESSES`` can be increased as in the example below:
+
+::
+
+    tethys gen asgi_service --asgi-processes <desired_number_of_instances> --conda-prefix <path_to_tethys_conda_environment>
+
+.. warning::
+
+    ``Bokeh Server`` does not currently support the use of multi intance processes. Therefore, it does not work with more than one ASGI process (``ASGI_PROCESSES`` must be equal to 1) (See `Bokeh Server documentation <https://docs.bokeh.org/en/latest/docs/reference/server/server.html#bokeh.server.server.Server>`_).
```

Comment:
Wow. Bokeh doesn't support multiple ASGI processes? Seems like a major limitation.

Reply:
Yes, after trying to make it right with the production layer, it was still not working so I decided to try our dam inventory app websocket branch and it worked. After that, I did some searching and it turns out that Bokeh doesn't support multi instances. This seems to be more of a Tornado limitation so maybe there is a way to do it since we a using channels, but either way that is on their side. I was debating about creating an issue about it on their repo.